### DISCUSSION TITLE
Support horovod on Hadoop Yarn

### DISCUSSION THREAD

**Discussion by zuston** [horovod/horovod#2785] [Q&A]
**Your question:**
Horovod support running on Hadoop Yarn directly, not using Spark?

**Comment by tgaddair** [comment]
Hey @zuston, there are some projects that allow you to run [MPI on Yarn](https://github.com/alibaba/mpich2-yarn). With that, you can run Horovod from within the MPI job.

**Comment by zuston** [comment]
Thanks your reply.

**Comment by zuston** [comment]
Reopen it. 
I have some questions about Horovod integration with Hadoop Yarn.

Now we use the[ linkedin/TonY](https://github.com/linkedin/TonY/) to support Tensorflow/PyTorch running on Yarn and want to integrate Horovod with it.
I found that Horovod supports horovodrun to start multiple process on remote server through ssh. 

However it's hard to support ssh on Yarn. As it's designed that TonY support starting multiple process by its worker not a coordinator start them (like horovodrun).

So can you give me some advise to support it? like how to implement a custom horovod runner?

I'm so glad to contribute Horovod to integrate with Hadoop Yarn using TonY.

@tgaddair looking forward to your reply.

**Comment by tgaddair** [comment]
Hey @zuston, this sounds great!

I think we can definitely find a solution that will work for TonY. I'm not very familiar with the architecture, but based on my understanding from [these](https://www.slideshare.net/ssuser72f42a/scaling-deep-learning-on-hadoop-at-linkedin) slides, it looks like you have an Application Master that starts the Task Executor processes, somewhat similar to Spark.

One option would be to take inspiration from [Horovod on Spark](https://github.com/horovod/horovod/blob/master/horovod/spark/runner.py#L47), where each Task Executor launches a Horovod Service that can receive commands to execute from the `horovodrun` coordinator running on the Application Master. This approach assumes that the Application Master is responsible for initiating the training process.

A second option is to have the workers start the training process independently, without any environment variables set, but then request this information from a "rendezvous server" running within the Application Master that can provide this information. This approach would require using Gloo as the control plane instead of MPI (because MPI workers must be started from the coordinator).

More concretely, you can run the [RendezvousServer](https://github.com/horovod/horovod/blob/master/horovod/runner/http/http_server.py#L174) on the Application Master, responsible for telling the workers what their rank is and peer IP addresses area. Then each worker, when they call `hvd.init()` will connect to this server to obtain this information before beginning the training [here](https://github.com/horovod/horovod/blob/master/horovod/common/gloo/gloo_context.cc#L63). This acts as a synchronization point for all the workers.

The only thing you need to provide to the workers when they start, in this case, is the URL of the RendezvousServer running on the Application Master. We would also need to modify the RendezvousServer to provide the rank info, which is currently only done when running Elastic Horovod [here](https://github.com/horovod/horovod/blob/master/horovod/common/gloo/gloo_context.cc#L155), but this is not too difficult to change.

Hope this makes sense, happy to have a call to discuss some of this in detail and learn more about your requirements.

**Comment by zuston** [comment]
Thanks for your reply @tgaddair.
According to your advice, i think the first option will work and also can implement it on TonY.

> More concretely, you can run the RendezvousServer on the Application Master, responsible for telling the workers what their rank is and peer IP addresses area. Then each worker, when they call hvd.init() will connect to this server to obtain this information before beginning the training here. This acts as a synchronization point for all the workers.

According to local multiple process test, i found actually horovod.init will build communication by env var. Right?
That means that i just need to export some environment variables when exec `python training.py` command on TonY workers. These env vars are like `HOROVOD_LOCAL_RANK=1`, `HOROVOD_GLOO_RENDEZVOUS_ADDR`. Right?

Besides, after reading  the [elastic horovod design doc](https://docs.google.com/document/d/15ZoHA5AeSI_boeyIBapg9WPXKrYXMRvPytPzQWTCTn4/edit?spm=a2c6h.12873639.0.0.12b934ac7AvS8Z), i know some Horovod architecture. But I think I need to understand Horovod more deeply, do you have relevant Horovod architecture design doc?

**Comment by tgaddair** [comment]
Hey @zuston, I'm actually working on a Horovod architecture doc I hope to have ready by the end of this week. I'll be sure to share it with you when it's ready.

The Contributors guide has a really good overview of the Horovod on Spark architecture [here](https://github.com/horovod/horovod/blob/master/docs/contributors.rst#horovod-on-spark), which might be of particular interest to you.

Yes, the important thing is that the workers are created with the correct environment variables defined, and can reach the rendezvous server (when using Gloo).

**Comment by zuston** [comment]
Oh. It's nice, looking forward to your doc!

**Comment by oliverhu** [comment]
Thanks @zuston for contributing Horovod support to TonY!

@tgaddair could you help review https://github.com/linkedin/TonY/pull/524/files as well?
